### DISCUSSION TITLE
Performing inference with oxigraph

### DISCUSSION THREAD

**Discussion by cravishe** [oxigraph/oxigraph#401] [Q&A]
I have been exploring oxigraph and would like to know steps to perform inference on stored graph data. Any guide on this will be extremely helpful.

**Comment by Tpt** [comment]
@chiru091096 Hi! Thank you for your interest in Oxigraph. Sadly, Oxigraph has no inference/reasoning system built-in yet (see #130). However, it supports [SPARQL update](https://www.w3.org/TR/sparql11-update/) and you can use it to infer new triples from the existing ones. For example `rdfs:subClassOf` transitive closure materialization could be written
```sparql
INSERT { ?s rdfs:subClassOf ?o } WHERE { ?s rdfs:subClassOf+ ?o }
```
You might have to run the update again each time you add new data to Oxigraph.
Please refer to the SPARQL related documentation if you want to know more about how to run SPARQL updates.

**Reply by harshakhmk** [reply]
> @chiru091096 Hi! Thank you for your interest in Oxigraph. Sadly, Oxigraph has no inference/reasoning system built-in yet (see #130). However, it supports [SPARQL update](https://www.w3.org/TR/sparql11-update/) and you can use it to infer new triples from the existing ones. For example `rdfs:subClassOf` transitive closure materialization could be written
> 
> ```sparql
> INSERT { ?s rdfs:subClassOf ?o } WHERE { ?s rdfs:subClassOf+ ?o }
> ```
> 
> You might have to run the update again each time you add new data to Oxigraph. Please refer to the SPARQL related documentation if you want to know more about how to run SPARQL updates.

I tried this, but getting "error: expected CONSTRUCT" and tried other queries for inferencing, but getting similar error

**Reply by Tpt** [reply]
Hi! It's because Oxigraph differentiates queries and updates, just like a lot of other SPARQL implementations. If you use the HTTP server you should use `/update` for updates and not `/query` that is only for read-only operations. If you use the Rust/Python/JS APIs, you should use the `update` method and not the `query` method.

**Comment by dsisnero** [comment]
Not sure if this would help but it parses owl ontologies https://github.com/phillord/horned-owl

**Comment by hobofan** [comment]
I'd recommend the following paper (which AFAIK still underpins Stardog's inference) to anyone trying to implement reasoning on larger datasets in oxigraph: https://www.cs.ox.ac.uk/people/george.konstantinidis/EvaluationOfOWL2.pdf (The Stardog documentation page on inference may also give some further hints on what to look out for in terms of limitations).

With that, as long as you keep up-to-date with the TBox (which can also be done in an incremental fashion), it can work quite well with a lot of often changing data, and without the need to materialize the inferred triples up-front.

**Comment by cravishe** [comment]
Thank you for quick support on this doubt. It is very helpful

**Comment by niklasl** [comment]
In case it might help someone: I recently transcribed (most of) OWL RL into a SPARQL update query: [Owlery](https://github.com/niklasl/owlery). It's just a PoC, with some serious limits, and _not_ how you're supposed to implement OWL. But depending on your use case, it might be useful. (For one, it seeks to show how simple the core of OWL reasoning is.)

It supports `owl:hasSelf` for rolification (that I missed in [reasonable](https://github.com/gtfierro/reasonable/)) and the repo contains some (WIP) extra updates for "generalization", RDF 1.2 triple terms, and `xsd:dateTime` "introspection". And (of course) a simple wrapper Python script using [pyoxigraph](https://pyoxigraph.readthedocs.io/).
